Add Beat config library

david-kow opened the pull request:
This PR introduces a configuration library for Beat CRD. Yaml files in `config/recipes/beats` are self-contained and should be fully functional in non-secured cluster (PSP/SCC is out of scope).


```
$ kubectl apply -f config/recipes/beats # due to conflicting naming not all examples will be deployed
```

```
$ kubectl get es,kb,apm,ent,beat -A
NAMESPACE   NAME                                                       HEALTH   NODES   VERSION   PHASE   AGE
default     elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch   green    3       7.8.0     Ready   2m39s

NAMESPACE   NAME                                  HEALTH   NODES   VERSION   AGE
default     kibana.kibana.k8s.elastic.co/kibana   green    1       7.8.0     2m39s

NAMESPACE   NAME                                   HEALTH   AVAILABLE   EXPECTED   TYPE          VERSION   AGE
default     beat.beat.k8s.elastic.co/auditbeat     green    3           3          auditbeat     7.8.0     2m39s
default     beat.beat.k8s.elastic.co/filebeat      green    3           3          filebeat      7.8.0     2m39s
default     beat.beat.k8s.elastic.co/heartbeat     green    1           1          heartbeat     7.8.0     2m38s
default     beat.beat.k8s.elastic.co/journalbeat   green    3           3          journalbeat   7.8.0     2m38s
default     beat.beat.k8s.elastic.co/metricbeat    green    1           1          metricbeat    7.8.0     2m38s
default     beat.beat.k8s.elastic.co/packetbeat    green    3           3          packetbeat    7.8.0     2m37s
```


Each recipe is tested as an E2E test with custom (per recipe) setup and checks. Docs linking to those recipes will come as a separate PR. Related: https://github.com/elastic/cloud-on-k8s/issues/3319.

barkbay commented on config/recipes/beats/filebeat_no_autodiscover.yaml:
```diff
@@ -0,0 +1,71 @@
+        securityContext:
+          runAsUser: 0
+          # If using Red Hat OpenShift uncomment this:
+          #privileged: true
```
The `privileged` field does not exist in the [`PodSecurityContext`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L3166). It is only available in the container's [`SecurityContext`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L5848).

Also I'm wondering if it is required for all versions of Openshift since it's seems to work out of the box on Openshift 4 according to our e2e tests ? I *think* it is required on Openshift 3.11 to bypass SELinux on the `hostPath` volume.

david-kow replied:
I've fixed the placement, I'll try to test at some point to see if only 3.11 is affected. FYI, the comment was taken from Beats examples, but you are right, it works in our E2E tests.

pebrc commented on config/recipes/beats/filebeat_autodiscover_by_metadata.yaml:
```diff
@@ -0,0 +1,115 @@
+apiVersion: beat.k8s.elastic.co/v1beta1
```
I feel like one or two sentences explaining what each config does would be helpful, especially for those of us who are not experts in the details of Beats configuration 

Maybe either have a subdirectory with README for each or alternatively a shared README with short explanation for each wdyt?

pebrc commented on test/e2e/beat/recipes_test.go:
```diff
@@ -0,0 +1,197 @@
+	customize := func(builder beat.Builder) beat.Builder {
+		return builder.
+			WithRoles(beat.PSPClusterRoleName, beat.AutodiscoverClusterRoleName).
+			WithESValidations(
```
I think it's a bit odd that we define test validations as part of the builder, that builds the test subjects, but I guess this is a general problem in our e2e setup at the moment.

pebrc commented on test/e2e/test/helper/yaml.go:
```diff
@@ -104,3 +108,212 @@ func (yd *YAMLDecoder) ToBuilders(reader *bufio.Reader, transform BuilderTransfo
+	return objects, nil
+}
+
+// RunFile runs builder workflow for all known resources in a yaml file, all other objects are created before and deleted
```
```suggestion
// RunFile runs the builder workflow for all known resources in a yaml file, all other objects are created before and deleted
```
"builder workflow" made me think twice to understand what you mean

pebrc commented on test/e2e/test/helper/yaml.go:
```diff
@@ -104,3 +108,212 @@ func (yd *YAMLDecoder) ToBuilders(reader *bufio.Reader, transform BuilderTransfo
+		}
+}
+
+func transform(namespace, fullTestName, suffix string, transformers []BuilderTransform, objects []runtime.Object) ([]test.Builder, []runtime.Object) {
```
`transformToE2E`  to be a bit more specific maybe?

pebrc commented on config/recipes/beats/packetbeat_dns_http.yaml:
```diff
@@ -0,0 +1,69 @@
+        - ip_port:
+        matchers:
+        - field_format:
+            format: '%{[ip]}:%{[port]}'
```
```
 E0714 12:34:56.169527       1 reflector.go:125] github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:146: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:default" cannot  │
│ list resource "pods" in API group "" at the cluster scope       
```

david-kow commented on config/recipes/beats/auditbeat_hosts.yaml:
```diff
@@ -0,0 +1,133 @@
+        indexers:
+        - container:
+        matchers:
+        - fields.lookup_fields: ['container.id']
```
Yes.

david-kow commented:
I went and added missing RBAC, but then I realized that those kubernetes processors are not adding anything to the events, so I removed them. 

In case of auditbeat and journalbeat I'm not sure if kubernetes metadata makes sense at all as they interact with hosts. For packetbeat, I think it would make sense to annotate events with src/dst containers, but it doesn't seem to happen and I'm not sure if this is a Beat limitation or something is wrong with the config. I think that I'd leave that for a future improvement.

anyasabo commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+
+#### Filebeat with autodiscover - `filebeat_autodiscover.yaml`
+
+Deploys Filebeat as DaemonSet with autodiscover feature enabled. All pods in all namespace will have logs shipped to a Elasticsearch cluster.
```
```suggestion
Deploys Filebeat as DaemonSet with autodiscover feature enabled. All pods in all namespaces will have logs shipped to a Elasticsearch cluster.
```

anyasabo commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+
+#### Metricbeat for Kubernetes monitoring - `metricbeat_hosts.yaml`
+
+Deploys Metricbeat as a DaemonSet that monitors the host resource usage (cpu, memory, network, filesystem) and Kubernetes resources (nodes, pods, containers, volumes).
```
```suggestion
Deploys Metricbeat as a DaemonSet that monitors the host resource usage (cpu, memory, network, filesystem) and Kubernetes resources (Nodes, Pods, Contains, Volumes).
```

anyasabo commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+
+#### Filebeat with autodiscover for metadata - `filebeat_autodiscover_by_metadata.yaml`
+
+Deploys Filebeat as a DaemonSet with autodiscover feature enabled. Fullfilling any of the two conditions below will cause a given Pod logs to be shipped to a Elasticsearch cluster:
```
```suggestion
Deploys Filebeat as a DaemonSet with autodiscover feature enabled. Fullfilling any of the two conditions below will cause a given Pod's logs to be shipped to an Elasticsearch cluster:
```

anyasabo commented on config/recipes/beats/filebeat_autodiscover_by_metadata.yaml:
```diff
@@ -0,0 +1,115 @@
+        serviceAccountName: filebeat
+        automountServiceAccountToken: true
+        terminationGracePeriodSeconds: 30
+        dnsPolicy: ClusterFirstWithHostNet
```
I realize this is in the Beats-provided examples, but it's not clear to me why filebeat would need access to the host network namespace. I didn't get a chance to test it without it yet though

david-kow replied:
I checked with @exekias and while it works without it, host metadata will be richer - `host.ip`, `host.mac` will be present. I'd be fine leaving it as it is.

anyasabo replied:
I think we at least want to leave a comment, I can't be the only one who sees this and it rings alarm bells. I'm also skeptical of how useful the host network info is for container logs.

anyasabo commented on config/recipes/beats/filebeat_autodiscover_by_metadata.yaml:
```diff
@@ -0,0 +1,115 @@
+  count: 1
+  elasticsearchRef:
+    name: elasticsearch
+---
```
```suggestion
...
```

`---` indicates the beginning of a yaml document and `...` the ending of one

anyasabo commented on config/recipes/beats/metricbeat_hosts.yaml:
```diff
@@ -0,0 +1,188 @@
+                system:
+                  filesystem:
+                    mount_point: ^/(sys|cgroup|proc|dev|etc|host|lib)($|/)
+      - module: kubernetes
```
Not necessary for this PR but I think it would be interesting to have this as a deployment example using autodiscover to scrape the kubelet metrics. There's not really any need to have this as a dset

david-kow replied:
Aren't these scoped to a node? I don't see all nodes/pods when I query kubelet endpoint.

anyasabo replied:
Correct, you could autodiscover each node and hit the kubelet on it

david-kow replied:
Hm, we could, but we already have a Metricbeat running everywhere for system metricsets, this would just add another one. I'll create an issue to discuss.

anyasabo commented:
>In case of auditbeat and journalbeat I'm not sure if kubernetes metadata makes sense at all as they interact with hosts.

k8s nodes still have labels and annotations, so I could see how it *might* happen, but going by your message it does not 🤷‍♀️ maybe worth opening a docs issue just to clarify what resources/beats actually add metadata

anyasabo commented:
> Yaml files in config/recipes/beats are self-contained

The three filebeat examples share a beat name so overlap each other, I was a bit confused at first when the k8s metadata was not being added.

anyasabo commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+# Beats Configuration Library
+
+This directory contains yaml manifests with example configurations for Beats. These manifests are self-contained and work out-of-the-box on any non-secured Kubernetes cluster. All of them contain three-node Elasticsearch cluster and single Kibana instance. All Beat configurations set up Kibana dashboards if they are available for a given Beat and all required RBAC resources. 
```
Should we add a note that none of these set resource requests/limits and they may want to set these in a production deployment?

sebgl replied:
> none of these set resource requests/limits

should we set default reasonable ones in the examples directly?

david-kow replied:
> > none of these set resource requests/limits
> 
> should we set default reasonable ones in the examples directly?

For Beats, I'd say our defaults are already reasonable. I think we shouldn't overload those examples.

anyasabo replied:
Ah I wondered because even on a cluster with very little traffic packetbeat was using up a lot of its CPU, but I'm good with not mentioning it

anyasabo commented:
I'm at least having trouble with the auditbeat recipe, the only events in the dashboard are 
>2020-07-14T22:11:16.890Z	ERROR	[auditd]	auditd/audit_linux.go:155	Failure receiving audit events	{"error": "failed to set audit PID. An audit process is already running (PID 152)"}

Even after I deleted the dset and let ECK recreate it. So hopped on and saw

>sabo@gke-sabo-dev-cluster-default-pool-c24f3bc2-ggb1 ~ $ ps aux | grep 152
root         152  0.0  0.8 340984 262828 ?       Ss   Jul08   0:35 /usr/lib/systemd/systemd-journald

And of course @jordansissel is forever in our hearts and already ran into the same issue:
https://github.com/elastic/beats/issues/8523

I wonder if we want to just link to that as a known COS issue. I can also see the argument for including the fix in a recipe though since we have a lot of GKE customers.

sebgl commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+# Beats Configuration Library
```
Can "library" be a confusing term to use? To me it sounds more like a programmatic thing than a collection of examples. I'd appreciate thought from better english speakers though (cc. @anyasabo @charith-elastic)

pebrc replied:
Good point! Maybe we should just go with "Beats Configuration Examples"?

charith-elastic replied:
_The Beats Zoo_?  :stuck_out_tongue: 
+1 for _Beats Configuration Examples_

anyasabo replied:
👍 Examples or Cookbook I think both avoid the ambiguity seb mentioned

sebgl commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+# Beats Configuration Library
+
+This directory contains yaml manifests with example configurations for Beats. These manifests are self-contained and work out-of-the-box on any non-secured Kubernetes cluster. All of them contain three-node Elasticsearch cluster and single Kibana instance. All Beat configurations set up Kibana dashboards if they are available for a given Beat and all required RBAC resources. 
```
Should we mention how the Elasticsearch spec is not suitable to a production environment?

david-kow replied:
This is already there in https://github.com/elastic/cloud-on-k8s/blob/master/config/recipes/README.asciidoc, should we just repeat it here?

sebgl replied:
I'm not sure people will notice the root README, it's worth repeating here imo.

sebgl commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+# Beats Configuration Library
```
I'm wondering how those examples will be referenced from the documentation on the website. I think there's value in listing all the examples explicitly in the docs (ie. copying this markdown file content, basically).

charith-elastic replied:
[I tried to include the samples in the documentation a long time ago](https://github.com/elastic/cloud-on-k8s/pull/2656) but the docs build only checks out the `docs` directory so that wasn't successful. We'll either have to move the samples inside the docs directory or figure out the magic incantation required by the docs build to checkout more directories.

david-kow replied:
I planned to copy it to the docs almost 1:1.

sebgl commented on config/recipes/beats/README.md:
```diff
@@ -0,0 +1,39 @@
+
+Deploys Journalbeat as a DaemonSet that ships data from systemd journals.
+
+
```
nit: new line

pebrc commented:
> Don't think this is a blocker but: I'm at least having trouble with the auditbeat recipe, the only events in the dashboard are

I forgot to mention I tested on AKS where I did not run into the auditbeat issue @anyasabo found.

david-kow commented:
> > Yaml files in config/recipes/beats are self-contained
> 
> The three filebeat examples share a beat name so overlap each other, I was a bit confused at first when the k8s metadata was not being added. Maybe this is fine, but we probably want to update the initial PR if so. Otherwise we may want to change the name of the beats

But is "self-contained" implying they won't overlap? I didn't mean to imply the latter. I'd be reluctant to rename, because we'd also need to rename RBAC pieces and ES/Kibana for consistency. Right now they are fairly easy to ready, but if we start encoding more things into the names they might get more complicated.

exekias commented:
@ChrsMark FYI

anyasabo commented on config/recipes/beats/heartbeat_es_kb_health.yaml:
```diff
@@ -0,0 +1,50 @@
+    replicas: 1
+    podTemplate:
+      spec:
+        dnsPolicy: ClusterFirstWithHostNet
```
Same here for heartbeat, I don't see the advantage of running in the host network and would break in clusters where the nodes can't communicate with the pods (albeit a rare situation). I think a comment justifying it makes sense

anyasabo commented:
>But is "self-contained" implying they won't overlap? I didn't mean to imply the latter.

I think that combined with the suggestion to just:

>kubectl apply -f config/recipes/beats

to review is what threw me off. I'm good with leaving the names as is, just a note for other reviewers 👍
